qos/evm/state.go

```diff
@@ -29,72 +31,164 @@ type ServiceState struct {
 			continue
 		}
 
-		// TODO_TECHDEBT: use a more resilient method for updating block height.
-		// E.g. one endpoint returning a very large number as block height should
-		// not result in all other endpoints being marked as invalid.
-		blockNumber, err := endpoint.GetBlockNumber()
+		// Retrieve the block number from the endpoint.
+		blockNumber, err := endpoint.getBlockNumber()
 		if err != nil {
 			logger.Info().Err(err).Msg("Skipping endpoint with invalid block number")
 			continue
 		}
 
+		// Update the perceived block number.
 		s.perceivedBlockNumber = blockNumber
 
-		logger.With("endpoint_block_number", blockNumber).Info().Msg("Updating latest block height")
+		// Attempt to retrieve the archival balance from the endpoint.
+		balance, err := endpoint.getArchivalBalance()
+		if err != nil {
+			logger.Info().Err(err).Msg("Skipping endpoint without archival balance")
+			continue
+		}
+
+		// Only count non-empty balances toward consensus.
+		if s.archivalState.balance == "" && balance != "" {
+			s.archivalState.balanceConsensus[balance]++
+		}
+	}
+
+	// Update archival block number and archival balance only if not yet set.
+	if s.perceivedBlockNumber != 0 && s.archivalState.blockNumberHex == "" {
+		s.assignArchivalBlockNumber()
+	}
+	if s.archivalState.balance == "" {
+		s.updateArchivalBalance(consensusThreshold)
 	}
 
 	return nil
 }
 
-// validateEndpointBlockNumber validates the supplied endpoint against the supplied perceived block number for the EVM blockchain.
-func validateEndpointBlockNumber(endpoint endpoint, perceivedBlockNumber uint64) error {
-	blockNumber, err := endpoint.GetBlockNumber()
-	if err != nil {
-		return err
+// assignArchivalBlockNumber returns a random archival block number based on the perceived block number.
+func (s *ServiceState) assignArchivalBlockNumber() string {
+	archivalThreshold := s.archivalCheckConfig.Threshold
+	minArchivalBlock := s.archivalCheckConfig.ContractStartBlock
+
+	var result string
+	if s.perceivedBlockNumber <= archivalThreshold {
+		result = blockNumberToHex(0)
+	} else {
+		maxBlockNumber := s.perceivedBlockNumber - archivalThreshold
+		if maxBlockNumber < minArchivalBlock {
+			result = blockNumberToHex(minArchivalBlock)
+		} else {
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+			rangeSize := maxBlockNumber - minArchivalBlock + 1
+			result = blockNumberToHex(minArchivalBlock + (r.Uint64() % rangeSize))
+		}
 	}
 
-	if blockNumber < perceivedBlockNumber {
-		return fmt.Errorf("endpoint has block height %d, perceived block height is %d", blockNumber, perceivedBlockNumber)
+	s.archivalState.blockNumberHex = result
+	return result
+}
+
+// updateArchivalBalance checks for consensus and updates the archival balance if it hasn't been set yet.
+func (s *ServiceState) updateArchivalBalance(consensusThreshold int) {
+	for balance, count := range s.archivalState.balanceConsensus {
+		if count >= consensusThreshold {
+			s.archivalState.balance = balance
```

Comment:
This actually breaks at the first balance that exceeds the threshold, but not necessarily the one with the highest count.

- If this is intentional, #PUC
- If this is not intentional, #PUC + TODO

Reply:
It's not checking balance but the number of endpoints that agree on the balance at the specific block height for the specific contract.